### ISSUE TITLE
Getting all combinations of characteristic-fraction-speciation-unit in harmonization table

### ISSUE THREAD

**Issue by ehinman** [USEPA/TADA#319]
**Describe the bug**

We want to make the harmonization table as complete as possible for nutrients and priority parameters, so that we are providing users with the most support in terms of synonyms and harmonizing data. We used a char-frac-spec combination spreadsheet Kevin pulled from WQX to see the most common combinations of all three, but I created a test for checking to make sure we weren't missing combos using random datasets and found many new combinations that weren't in the WQX spreadsheet. At first, I thought this was because WQX does not account for NWIS char-frac-spec combinations, but I realized that this might be more extensive than just the NWIS data stream: the WQX combinations (and the WQX char validation table) do not consider blanks or NA's in any of the columns. Thus, for characteristics for which a fraction and speciation are NOT required, we would need to separately add all of those combinations to the harmonization template. 

**To Reproduce**

Code to reproduce the behavior:

``` r
remotes::install_github("USEPA/TADA", ref = "develop")
library(TADA)

 test = TADA_RandomTestingSet()
  test1 = TADA_RunKeyFlagFunctions(test)
  ref = TADA_GetSynonymRef()
  ref_chars = unique(ref$TADA.CharacteristicName)
  test_chars = unique(subset(test1, test1$TADA.CharacteristicName%in%ref_chars)[,c("TADA.CharacteristicName","TADA.ResultSampleFractionText","TADA.MethodSpecificationName","TADA.ResultMeasure.MeasureUnitCode")])
  test_chars_ref = merge(test_chars, ref, all.x = TRUE)
  new_combos = subset(test_chars_ref, is.na(test_chars_ref$HarmonizationGroup))[,c("TADA.CharacteristicName","TADA.ResultSampleFractionText","TADA.MethodSpecificationName","TADA.ResultMeasure.MeasureUnitCode")]
  if(dim(new_combos)[1]>0){
    print("New combinations found in random dataset test:")
    print(new_combos)
  }
```

**Expected behavior**

Ideally, we could pull all combinations of these characteristics from the water quality PORTAL, where NWIS and WQX mix with all allowable values. 

**Reminders for TADA contributors addressing this issue**

New features should include all of the following work:

-   [ ] Create the function/code.

-   [ ] Document all code using comments to describe what is does.

-   [ ] Create tests in tests folder.

-   [ ] Create help file using roxygen2 above code.

-   [ ] Create working examples in help file (via roxygen2).

-   [ ] Add to appropriate vignette (or create new one).

**Comment by cristinamullin** [comment]
This won't fix this issue, but it would help reduce total combinations. We could change all NONE to NA for speciation and fraction as part of autoclean, assuming NONE is equivalent to NA.

Cristina A Mullin, PhD (she/her)
Water Data Integration Branch
Watershed Restoration, Assessment, and Protection Division
Office of Wetlands, Oceans and Watersheds
US EPA|Office of Water
***@***.******@***.***>

From: Elise H. ***@***.***>
Sent: Friday, August 4, 2023 5:32 PM
To: USEPA/TADA ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [USEPA/TADA] Getting all combinations of characteristic-fraction-speciation-unit in harmonization table (Issue #319)


Describe the bug

We want to make the harmonization table as complete as possible for nutrients and priority parameters, so that we are providing users with the most support in terms of synonyms and harmonizing data. We used a char-frac-spec combination spreadsheet Kevin pulled from WQX to see the most common combinations of all three, but I created a test for checking to make sure we weren't missing combos using random datasets and found many new combinations that weren't in the WQX spreadsheet. At first, I thought this was because WQX does not account for NWIS char-frac-spec combinations, but I realized that this might be more extensive than just the NWIS data stream: the WQX combinations (and the WQX char validation table) do not consider blanks or NA's in any of the columns. Thus, for characteristics for which a fraction and speciation are NOT required, we would need to separately add all of those combinations to the harmonization template.

To Reproduce

Code to reproduce the behavior:

remotes::install_github("USEPA/TADA", ref = "develop")

library(TADA)



 test = TADA_RandomTestingSet()

  test1 = TADA_RunKeyFlagFunctions(test)

  ref = TADA_GetSynonymRef()

  ref_chars = unique(ref$TADA.CharacteristicName)

  test_chars = unique(subset(test1, test1$TADA.CharacteristicName%in%ref_chars)[,c("TADA.CharacteristicName","TADA.ResultSampleFractionText","TADA.MethodSpecificationName","TADA.ResultMeasure.MeasureUnitCode")])

  test_chars_ref = merge(test_chars, ref, all.x = TRUE)

  new_combos = subset(test_chars_ref, is.na(test_chars_ref$HarmonizationGroup))[,c("TADA.CharacteristicName","TADA.ResultSampleFractionText","TADA.MethodSpecificationName","TADA.ResultMeasure.MeasureUnitCode")]

  if(dim(new_combos)[1]>0){

    print("New combinations found in random dataset test:")

    print(new_combos)

  }

Expected behavior

Ideally, we could pull all combinations of these characteristics from the water quality PORTAL, where NWIS and WQX mix with all allowable values.

Reminders for TADA contributors addressing this issue

New features should include all of the following work:

  *   [ ] Create the function/code.
  *   [ ] Document all code using comments to describe what is does.
  *   [ ] Create tests in tests folder.
  *   [ ] Create help file using roxygen2 above code.
  *   [ ] Create working examples in help file (via roxygen2).
  *   [ ] Add to appropriate vignette (or create new one).

-

**Comment by ehinman** [comment]
We could also change all NONEs in the validation table to NA, too. That might be a solid solution.

**Comment by cristinamullin** [comment]
That would be my preference, to change NONE to NA in autoclean and in the reference table.

**Comment by ehinman** [comment]
Ok, got it. Sounds good. Thanks!

**Comment by ehinman** [comment]
TADA_Autoclean converts all NONE to NA in the fraction and speciation columns prior to validation, and the validation table has NONE set to INVALID. This still does not address additional combos coming in from NWIS (NONE vs NA aside).

**Comment by cristinamullin** [comment]
Note: TADA priority characteristics are here: https://usepa.sharepoint.com/:x:/r/sites/WQPDataAssessmentTeam/_layouts/15/Doc.aspx?sourcedoc=%7B65B3DD61-5856-48CD-8A3C-25415DE94955%7D&file=TADA_Supported_Characteristics.xlsx&action=default&mobileredirect=true

**Comment by hillarymarler** [comment]
Might be time to revisit this one?

**Comment by cristinamullin** [comment]
I agree, especially for priority characteristics. Maybe we can start with bacteria to support the work Kenny is doing with the TADA Module 3 subgroup. Note: if we add new nitrogen or phosphorus combinations, they also need to be added to the np summation key and prioritized (given a ranking) in the summation equations. I wonder if there is a simpler way...

**Comment by cristinamullin** [comment]
See related issue: https://github.com/USEPA/EPATADA/issues/79

**Comment by cristinamullin** [comment]
The NA/NONE issue has already been addressed and if new combinations are added to the harmonization template, there is a test to make sure they are also added to the TN/TP ref... I can create a test to look for new combinations but it is going to be really difficult to find them ALL
